lib/cipherstash/index/bloom_filter.rb

```diff
@@ -0,0 +1,87 @@
+require_relative "../client/error"
+
+module CipherStash
+  class Index
+    # A bloom filter implementation designed to be used with the *FilterMatch index classes.
+    #
+    # @private
+    class BloomFilter
+      FILTER_TERM_BITS_MIN = 3
+      FILTER_TERM_BITS_MAX = 16
+      FILTER_TERM_BITS_DEFAULT = 3
+      FILTER_SIZE_DEFAULT = 256
+      VALID_FILTER_SIZES = [32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536]
+
+      # The "set" bits of the bloom filter
+      attr_reader :bits
+
+      # The size of the bloom filter in bits. Commonly referred to as "m". Since we only keep track of the set bits, the filter
+      # size determines the maximum value of the positions stored in the bits attr.
+      #
+      # Valid values are powers of 2 from 32 to 65536.
+      #
+      # @return [Integer]
+      attr_reader :filter_size
+
+      # The number of bits to set per term. Commonly referred to as "k".
+      #
+      # Valid values are integers from 3 to 16.
+      #
+      # @return [Integer]
+      attr_reader :filter_term_bits
+
+      # Creates a new bloom filter with the given key and filter match index settings.
+      #
+      # @param key [String] the key to use for hashing terms.
+      #
+      # @param opts [Hash] the index settings.
+      #   "filterSize" and "filterTermBits" are used to set the filter_size and filter_term_bits attrs.
+      #
+      # @raise [CipherStash::Client::Error::InvalidSchemaError] if an invalid "filterSize" or "filterTermBits" is given.
+      def initialize(key, opts = {})
+        @key = key
+        @bits = Set.new()
+
+        @filter_size = opts["filterSize"] || FILTER_SIZE_DEFAULT
+
+        if not VALID_FILTER_SIZES.include?(@filter_size)
```

Comment:
Doesn't make a difference here but generally `!` is preferred to `not` as it has higher precedence. See https://stackoverflow.com/questions/27429711/difference-between-not-and-in-ruby

Reply:
Reading this code again, I could also use `unless` instead of `if not` or `if !`. Do Rubyists tend to prefer `unless` to `if !`/`if not`?

Reply:
Opinions most likely to result in bloodshed at a Ruby meetup, in order of likelihood:

1. Tabs vs Spaces
2. Vim vs Emacs (vs nvi vs Code vs...)
3. `if !` vs `unless`

*Personally*, I use `unless` rather than `if !`, when (and only when) there is a single predicate being tested and there is no `else` clause.  The latter is obvious (if you've got an else clause you can just invert the predicate and swap the `if` and `else` blocks), the former is for readability -- `unless foo && bar || !baz` is a lot harder to mentally reason about, IMO, than the `if`-based equivalent.

Reply:
Updated these checks to use `unless`.